File: repertoire/data-integration/providers/connector_protocol/CopfRemote.swift
// CopfRemote.swift — connector_protocol provider
// Connects to another Clef instance API for schema sharing, identity field mapping, and bidirectional sync

import Foundation

public let clefRemoteProviderId = "clef_remote"
public let clefRemotePluginType = "connector_protocol"

public struct CrConnectorConfig: Codable {
    public var baseUrl: String?; public var connectionString: String?
    public var auth: [String: String]?; public var headers: [String: String]?; public var options: [String: String]?
}
public struct CrQuerySpec: Codable {
    public var path: String?; public var query: String?; public var params: [String: String]?
    public var cursor: String?; public var limit: Int?
}
public struct CrWriteResult: Codable { public var created: Int; public var updated: Int; public var skipped: Int; public var errors: Int }
public struct CrTestResult: Codable { public var connected: Bool; public var message: String; public var latencyMs: Int? }
public struct CrStreamDef: Codable { public var name: String; public var schema: [String: String]; public var supportedSyncModes: [String] }
public struct CrDiscoveryResult: Codable { public var streams: [CrStreamDef] }
public enum CopfRemoteError: Error { case connectionFailed(String); case apiFailed(String) }

private struct CopfApiResponse: Codable {
    var success: Bool
    var data: AnyCodable?
    var error: String?
    var meta: CopfMeta?
}

private struct CopfMeta: Codable {
    var cursor: String?
    var hasMore: Bool?
    var total: Int?
}

private struct AnyCodable: Codable {
    let value: Any
    init(_ value: Any) { self.value = value }
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let arr = try? container.decode([[String: String]].self) { value = arr }
        else if let dict = try? container.decode([String: String].self) { value = dict }
        else { value = try container.decode(String.self) }
    }
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let str = value as? String { try container.encode(str) }
    }
}

private struct FieldMapping {
    var local: String; var remote: String; var transform: String?
}

private func buildApiUrl(_ baseUrl: String, path: String) -> String {
    "\(baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")))/api/v1\(path)"
}

private func buildHeaders(_ config: CrConnectorConfig) -> [String: String] {
    var headers = ["Content-Type": "application/json", "X-Clef-Client": "connector_protocol/clef_remote"]
    if let cfgH = config.headers { headers.merge(cfgH) { _, new in new } }
    if let token = config.auth?["token"] { headers["Authorization"] = "Bearer \(token)" }
    if let apiKey = config.auth?["apiKey"] { headers["X-Clef-API-Key"] = apiKey }
    return headers
}

private func getFieldMappings(_ config: CrConnectorConfig) -> [FieldMapping] {
    // Parse field mappings from options (format: "local:remote:transform,...")
    guard let raw = config.options?["fieldMappings"] else { return [] }
    return raw.split(separator: ",").compactMap { pair -> FieldMapping? in
        let parts = pair.split(separator: ":").map { String($0).trimmingCharacters(in: .whitespaces) }
        guard parts.count >= 2 else { return nil }
        return FieldMapping(local: parts[0], remote: parts[1], transform: parts.count > 2 ? parts[2] : nil)
    }
}

private func mapFields(_ record: [String: Any], mappings: [FieldMapping], toRemote: Bool) -> [String: Any] {
    if mappings.isEmpty { return record }
    var mapped = [String: Any]()
    let mappedSources = Set(mappings.map { toRemote ? $0.local : $0.remote })
    for mapping in mappings {
        let src = toRemote ? mapping.local : mapping.remote
        let tgt = toRemote ? mapping.remote : mapping.local
        if let value = record[src] {
            var transformed: Any = value
            switch mapping.transform {
            case "toString": transformed = "\(value)"
            case "toLowerCase": transformed = "\(value)".lowercased()
            case "toUpperCase": transformed = "\(value)".uppercased()
            default: break
            }
            mapped[tgt] = transformed
        }
    }
    for (key, value) in record where !mappedSources.contains(key) {
        mapped[key] = value
    }
    return mapped
}

public final class CopfRemoteConnectorProvider {
    private let session = URLSession(configuration: .default)
    public init() {}

    public func read(query: CrQuerySpec, config: CrConnectorConfig) async throws -> AsyncStream<[String: Any]> {
        let baseUrl = config.baseUrl ?? ""
        let headers = buildHeaders(config)
        let concept = query.path ?? ""
        let limit = query.limit ?? 100
        let mappings = getFieldMappings(config)

        return AsyncStream { continuation in
            Task {
                var cursor = query.cursor
                var hasMore = true
                var totalYielded = 0

                while hasMore && totalYielded < limit {
                    let pageSize = min(limit - totalYielded, 100)
                    var urlStr = buildApiUrl(baseUrl, path: "/concepts/\(concept)/records?limit=\(pageSize)")
                    if let c = cursor { urlStr += "&cursor=\(c)" }
                    if let params = query.params {
                        for (k, v) in params { urlStr += "&\(k)=\(v)" }
                    }
                    guard let url = URL(string: urlStr) else { break }
                    var request = URLRequest(url: url)
                    for (k, v) in headers { request.setValue(v, forHTTPHeaderField: k) }

                    do {
                        let (data, resp) = try await self.session.data(for: request)
                        guard let httpResp = resp as? HTTPURLResponse, httpResp.statusCode < 300 else { break }
                        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                              let success = json["success"] as? Bool, success,
                              let records = json["data"] as? [[String: Any]] else { break }

                        for record in records {
                            let mapped = mapFields(record, mappings: mappings, toRemote: false)
                            continuation.yield(mapped)
                            totalYielded += 1
                        }

                        let meta = json["meta"] as? [String: Any]
                        hasMore = (meta?["hasMore"] as? Bool) ?? false
                        cursor = meta?["cursor"] as? String
                    } catch { break }
                }
                continuation.finish()
            }
        }
    }

    public func write(records: [[String: Any]], config: CrConnectorConfig) async throws -> CrWriteResult {
        let baseUrl = config.baseUrl ?? ""
        let headers = buildHeaders(config)
        let concept = config.options?["concept"] ?? ""
        let syncMode = config.options?["syncMode"] ?? "upsert"
        let mappings = getFieldMappings(config)
        let batchSize = Int(config.options?["batchSize"] ?? "50") ?? 50
        var result = CrWriteResult(created: 0, updated: 0, skipped: 0, errors: 0)

        let mapped = records.map { mapFields($0, mappings: mappings, toRemote: true) }

        for i in stride(from: 0, to: mapped.count, by: batchSize) {
            let batch = Array(mapped[i..<min(i + batchSize, mapped.count)])
            let urlStr = buildApiUrl(baseUrl, path: "/concepts/\(concept)/sync")
            guard let url = URL(string: urlStr) else { continue }
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            for (k, v) in headers { request.setValue(v, forHTTPHeaderField: k) }
            let body: [String: Any] = ["records": batch, "mode": syncMode]
            request.httpBody = try? JSONSerialization.data(withJSONObject: body)

            do {
                let (data, resp) = try await session.data(for: request)
                if let httpResp = resp as? HTTPURLResponse, httpResp.statusCode < 300,
                   let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let respData = json["data"] as? [String: Any] {
                    result.created += (respData["created"] as? Int) ?? 0
                    result.updated += (respData["updated"] as? Int) ?? 0
                    result.skipped += (respData["skipped"] as? Int) ?? 0
                } else {
                    result.errors += batch.count
                }
            } catch { result.errors += batch.count }
        }
        return result
    }

    public func test(config: CrConnectorConfig) async throws -> CrTestResult {
        let baseUrl = config.baseUrl ?? ""
        let headers = buildHeaders(config)
        let start = Date()
        let urlStr = buildApiUrl(baseUrl, path: "/health")
        guard let url = URL(string: urlStr) else {
            return CrTestResult(connected: false, message: "Invalid URL", latencyMs: 0)
        }
        var request = URLRequest(url: url)
        for (k, v) in headers { request.setValue(v, forHTTPHeaderField: k) }
        do {
            let (data, resp) = try await session.data(for: request)
            let ms = Int(Date().timeIntervalSince(start) * 1000)
            guard let httpResp = resp as? HTTPURLResponse, httpResp.statusCode < 300 else {
                return CrTestResult(connected: false, message: "HTTP error", latencyMs: ms)
            }
            let json = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["data"] as? [String: Any]
            let instance = (json?["instance"] as? String) ?? "unknown"
            let version = (json?["version"] as? String) ?? "?"
            return CrTestResult(connected: true, message: "Connected to Clef instance: \(instance) (v\(version))", latencyMs: ms)
        } catch {
            let ms = Int(Date().timeIntervalSince(start) * 1000)
            return CrTestResult(connected: false, message: error.localizedDescription, latencyMs: ms)
        }
    }

    public func discover(config: CrConnectorConfig) async throws -> CrDiscoveryResult {
        let baseUrl = config.baseUrl ?? ""
        let headers = buildHeaders(config)
        let urlStr = buildApiUrl(baseUrl, path: "/concepts")
        guard let url = URL(string: urlStr) else { return CrDiscoveryResult(streams: []) }
        var request = URLRequest(url: url)
        for (k, v) in headers { request.setValue(v, forHTTPHeaderField: k) }
        do {
            let (data, resp) = try await session.data(for: request)
            guard let httpResp = resp as? HTTPURLResponse, httpResp.statusCode < 300,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let concepts = json["data"] as? [[String: Any]] else {
                return CrDiscoveryResult(streams: [])
            }
            return CrDiscoveryResult(streams: concepts.compactMap { c in
                guard let name = c["name"] as? String else { return nil }
                var schema = [String: String]()
                if let fields = c["fields"] as? [[String: Any]] {
                    for field in fields {
                        if let fname = field["name"] as? String, let ftype = field["type"] as? String {
                            schema[fname] = ftype
                        }
                    }
                }
                return CrStreamDef(name: name, schema: schema, supportedSyncModes: ["full_refresh", "incremental", "bidirectional"])
            })
        } catch { return CrDiscoveryResult(streams: []) }
    }
}

File: repertoire/data-integration/providers/connector_protocol/copf_remote.rs
// CopfRemote — connector_protocol provider
// Connects to another COPF instance API for schema sharing, identity field mapping, and bidirectional sync

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::time::Instant;

pub const PROVIDER_ID: &str = "copf_remote";
pub const PLUGIN_TYPE: &str = "connector_protocol";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectorConfig {
    pub base_url: Option<String>,
    pub connection_string: Option<String>,
    pub auth: Option<HashMap<String, String>>,
    pub headers: Option<HashMap<String, String>>,
    pub options: Option<HashMap<String, Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QuerySpec {
    pub path: Option<String>,
    pub query: Option<String>,
    pub params: Option<HashMap<String, Value>>,
    pub cursor: Option<String>,
    pub limit: Option<u64>,
}

pub type Record = HashMap<String, Value>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WriteResult { pub created: u64, pub updated: u64, pub skipped: u64, pub errors: u64 }
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TestResult { pub connected: bool, pub message: String, pub latency_ms: Option<u64> }
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StreamDef { pub name: String, pub schema: HashMap<String, Value>, pub supported_sync_modes: Vec<String> }
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscoveryResult { pub streams: Vec<StreamDef> }

#[derive(Debug)]
pub struct ConnectorError(pub String);
impl std::fmt::Display for ConnectorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { write!(f, "{}", self.0) }
}
impl std::error::Error for ConnectorError {}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct CopfApiResponse {
    success: bool,
    data: Option<Value>,
    error: Option<String>,
    meta: Option<CopfMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct CopfMeta {
    cursor: Option<String>,
    has_more: Option<bool>,
    total: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct FieldMapping {
    local: String,
    remote: String,
    transform: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ConceptSchema {
    name: String,
    identity_fields: Vec<String>,
    fields: Vec<ConceptField>,
    conventions: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ConceptField {
    name: String,
    #[serde(rename = "type")]
    field_type: String,
    required: bool,
}

fn build_api_url(base_url: &str, path: &str) -> String {
    format!("{}/api/v1{}", base_url.trim_end_matches('/'), path)
}

fn build_headers(config: &ConnectorConfig) -> HashMap<String, String> {
    let mut headers = HashMap::new();
    headers.insert("Content-Type".into(), "application/json".into());
    headers.insert("X-COPF-Client".into(), "connector_protocol/copf_remote".into());
    if let Some(cfg_h) = &config.headers {
        headers.extend(cfg_h.clone());
    }
    if let Some(auth) = &config.auth {
        if let Some(token) = auth.get("token") {
            headers.insert("Authorization".into(), format!("Bearer {}", token));
        }
        if let Some(api_key) = auth.get("apiKey") {
            headers.insert("X-COPF-API-Key".into(), api_key.clone());
        }
    }
    headers
}

fn get_field_mappings(config: &ConnectorConfig) -> Vec<FieldMapping> {
    config.options.as_ref()
        .and_then(|o| o.get("fieldMappings"))
        .and_then(|v| serde_json::from_value(v.clone()).ok())
        .unwrap_or_default()
}

fn map_fields(record: &Record, mappings: &[FieldMapping], to_remote: bool) -> Record {
    if mappings.is_empty() { return record.clone(); }
    let mut mapped = Record::new();
    for mapping in mappings {
        let (src, tgt) = if to_remote { (&mapping.local, &mapping.remote) } else { (&mapping.remote, &mapping.local) };
        if let Some(value) = record.get(src) {
            let transformed = match mapping.transform.as_deref() {
                Some("toString") => json!(value.to_string()),
                Some("toLowerCase") => json!(value.as_str().map(|s| s.to_lowercase()).unwrap_or_default()),
                Some("toUpperCase") => json!(value.as_str().map(|s| s.to_uppercase()).unwrap_or_default()),
                _ => value.clone(),
            };
            mapped.insert(tgt.clone(), transformed);
        }
    }
    // Include unmapped fields
    let mapped_sources: Vec<&str> = mappings.iter()
        .map(|m| if to_remote { m.local.as_str() } else { m.remote.as_str() })
        .collect();
    for (key, value) in record {
        if !mapped_sources.contains(&key.as_str()) {
            mapped.insert(key.clone(), value.clone());
        }
    }
    mapped
}

pub struct CopfRemoteConnectorProvider {
    config: Option<ConnectorConfig>,
    client: reqwest::Client,
}

impl CopfRemoteConnectorProvider {
    pub fn new() -> Self {
        Self { config: None, client: reqwest::Client::new() }
    }

    pub async fn read(&self, query: &QuerySpec, config: &ConnectorConfig) -> Result<Vec<Record>, ConnectorError> {
        let base_url = config.base_url.as_deref().unwrap_or("");
        let headers = build_headers(config);
        let concept = query.path.as_deref().unwrap_or("");
        let limit = query.limit.unwrap_or(100);
        let field_mappings = get_field_mappings(config);

        let mut all_records = Vec::new();
        let mut cursor = query.cursor.clone();
        let mut has_more = true;

        while has_more && (all_records.len() as u64) < limit {
            let remaining = limit - all_records.len() as u64;
            let page_size = std::cmp::min(remaining, 100);
            let mut url = build_api_url(base_url, &format!("/concepts/{}/records?limit={}", concept, page_size));
            if let Some(c) = &cursor { url.push_str(&format!("&cursor={}", c)); }
            if let Some(params) = &query.params {
                for (k, v) in params { url.push_str(&format!("&{}={}", k, v)); }
            }

            let mut req = self.client.get(&url);
            for (k, v) in &headers { req = req.header(k.as_str(), v.as_str()); }

            let resp = req.send().await.map_err(|e| ConnectorError(e.to_string()))?;
            let result: CopfApiResponse = resp.json().await.map_err(|e| ConnectorError(e.to_string()))?;

            if !result.success { break; }
            if let Some(data) = result.data {
                if let Some(arr) = data.as_array() {
                    for item in arr {
                        if let Ok(rec) = serde_json::from_value::<Record>(item.clone()) {
                            all_records.push(map_fields(&rec, &field_mappings, false));
                        }
                    }
                }
            }

            has_more = result.meta.as_ref().and_then(|m| m.has_more).unwrap_or(false);
            cursor = result.meta.as_ref().and_then(|m| m.cursor.clone());
        }
        Ok(all_records)
    }

    pub async fn write(&self, records: &[Record], config: &ConnectorConfig) -> Result<WriteResult, ConnectorError> {
        let base_url = config.base_url.as_deref().unwrap_or("");
        let headers = build_headers(config);
        let concept = config.options.as_ref().and_then(|o| o.get("concept")).and_then(|v| v.as_str()).unwrap_or("");
        let sync_mode = config.options.as_ref().and_then(|o| o.get("syncMode")).and_then(|v| v.as_str()).unwrap_or("upsert");
        let field_mappings = get_field_mappings(config);
        let batch_size = config.options.as_ref().and_then(|o| o.get("batchSize")).and_then(|v| v.as_u64()).unwrap_or(50) as usize;

        let mut result = WriteResult { created: 0, updated: 0, skipped: 0, errors: 0 };
        let mapped: Vec<Record> = records.iter().map(|r| map_fields(r, &field_mappings, true)).collect();

        for chunk in mapped.chunks(batch_size) {
            let url = build_api_url(base_url, &format!("/concepts/{}/sync", concept));
            let body = json!({ "records": chunk, "mode": sync_mode });
            let mut req = self.client.post(&url);
            for (k, v) in &headers { req = req.header(k.as_str(), v.as_str()); }

            match req.json(&body).send().await {
                Ok(resp) if resp.status().is_success() => {
                    if let Ok(api_resp) = resp.json::<CopfApiResponse>().await {
                        if let Some(data) = api_resp.data {
                            result.created += data.get("created").and_then(|v| v.as_u64()).unwrap_or(0);
                            result.updated += data.get("updated").and_then(|v| v.as_u64()).unwrap_or(0);
                            result.skipped += data.get("skipped").and_then(|v| v.as_u64()).unwrap_or(0);
                        }
                    }
                }
                _ => result.errors += chunk.len() as u64,
            }
        }
        Ok(result)
    }

    pub async fn test_connection(&self, config: &ConnectorConfig) -> Result<TestResult, ConnectorError> {
        let base_url = config.base_url.as_deref().unwrap_or("");
        let headers = build_headers(config);
        let start = Instant::now();
        let url = build_api_url(base_url, "/health");

        let mut req = self.client.get(&url);
        for (k, v) in &headers { req = req.header(k.as_str(), v.as_str()); }

        match req.send().await {
            Ok(resp) if resp.status().is_success() => {
                let body: CopfApiResponse = resp.json().await.map_err(|e| ConnectorError(e.to_string()))?;
                let instance = body.data.as_ref().and_then(|d| d.get("instance")).and_then(|v| v.as_str()).unwrap_or("unknown");
                let version = body.data.as_ref().and_then(|d| d.get("version")).and_then(|v| v.as_str()).unwrap_or("?");
                Ok(TestResult {
                    connected: true,
                    message: format!("Connected to COPF instance: {} (v{})", instance, version),
                    latency_ms: Some(start.elapsed().as_millis() as u64),
                })
            }
            Ok(resp) => Ok(TestResult { connected: false, message: format!("HTTP {}", resp.status()), latency_ms: Some(start.elapsed().as_millis() as u64) }),
            Err(e) => Ok(TestResult { connected: false, message: e.to_string(), latency_ms: Some(start.elapsed().as_millis() as u64) }),
        }
    }

    pub async fn discover(&self, config: &ConnectorConfig) -> Result<DiscoveryResult, ConnectorError> {
        let base_url = config.base_url.as_deref().unwrap_or("");
        let headers = build_headers(config);
        let url = build_api_url(base_url, "/concepts");

        let mut req = self.client.get(&url);
        for (k, v) in &headers { req = req.header(k.as_str(), v.as_str()); }

        match req.send().await {
            Ok(resp) if resp.status().is_success() => {
                let body: CopfApiResponse = resp.json().await.map_err(|e| ConnectorError(e.to_string()))?;
                let concepts: Vec<ConceptSchema> = body.data.and_then(|d| serde_json::from_value(d).ok()).unwrap_or_default();
                Ok(DiscoveryResult {
                    streams: concepts.into_iter().map(|c| {
                        let mut schema = HashMap::new();
                        schema.insert("type".into(), json!("object"));
                        let props: HashMap<String, Value> = c.fields.iter()
                            .map(|f| (f.name.clone(), json!({"type": f.field_type, "required": f.required})))
                            .collect();
                        schema.insert("properties".into(), serde_json::to_value(props).unwrap_or_default());
                        schema.insert("identityFields".into(), json!(c.identity_fields));
                        schema.insert("conventions".into(), json!(c.conventions));
                        StreamDef {
                            name: c.name,
                            schema,
                            supported_sync_modes: vec!["full_refresh".into(), "incremental".into(), "bidirectional".into()],
                        }
                    }).collect(),
                })
            }
            _ => Ok(DiscoveryResult { streams: Vec::new() }),
        }
    }
}
